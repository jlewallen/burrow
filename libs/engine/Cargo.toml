[package]
name = "engine"
version = "0.1.0"
authors = ["Jacob Lewallen <jlewallen@gmail.com>"]
edition = "2021"

[lib]
doctest = false

[dependencies.kernel]
path = "../kernel"

[dependencies]
anyhow = "1.0.72"
chrono = { version = "0.4.26", features = ["serde"] }
serde = { version = "1.0.180", features = ["derive"] }
serde_json = "1.0.104"
tracing = "0.1.37"
tracing-subscriber = { version = "0.3.17", features = ["json"] }
