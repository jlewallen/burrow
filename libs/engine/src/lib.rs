mod identifiers;
mod internal;
mod perform;
mod users;

pub mod domain;
pub mod notifications;
pub mod sequences;
pub mod session;
pub mod storage;

pub use domain::*;
pub use notifications::*;
pub use session::*;

pub use users::model::{add_username_to_key, username_to_key};
