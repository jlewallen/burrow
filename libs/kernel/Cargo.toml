[package]
name = "kernel"
version = "0.1.0"
authors = ["Jacob Lewallen <jlewallen@gmail.com>"]
edition = "2021"

[lib]
doctest = false

[dependencies.replies]
path = "../replies"

[dependencies]
anyhow = "1.0.71"
thiserror = "1.0.40"
tracing = "0.1.37"
nom = "7.1.3"
serde_json = "1.0.96"
serde = { version = "1.0.163", features = ["derive"] }
treediff = { version = "4.0.2", features = ["with-serde-json"] }