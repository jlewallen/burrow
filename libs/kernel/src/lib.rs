pub mod actions;
pub mod hooks;
pub mod model;
pub mod perms;
pub mod plugins;
pub mod session;
pub mod surround;

pub mod common {
    pub use replies::*;
}

pub mod prelude {
    pub use replies::DomainEvent;

    pub use crate::actions::*;
    pub use crate::finder::*;
    pub use crate::hooks::*;
    pub use crate::model::compare::*;
    pub use crate::model::*;
    pub use crate::plugins::*;
    pub use crate::session::*;
    pub use crate::surround::*;
}

mod finder {
    use crate::model::{Audience, EntityKey, Entry, Item};
    use crate::surround::Surroundings;

    pub trait Finder: Send + Sync {
        fn find_world(&self) -> anyhow::Result<Entry>;

        fn find_location(&self, entry: &Entry) -> anyhow::Result<Entry>;

        fn find_item(
            &self,
            surroundings: &Surroundings,
            item: &Item,
        ) -> anyhow::Result<Option<Entry>>;

        fn find_audience(&self, audience: &Audience) -> anyhow::Result<Vec<EntityKey>>;
    }
}

#[macro_export]
macro_rules! here {
    () => {{
        format!("{}:{}", file!(), line!())
    }};
}

#[cfg(test)]
mod tests;
