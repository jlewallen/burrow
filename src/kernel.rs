use anyhow::Result;
use markdown_gen::markdown;
use once_cell::sync::Lazy;
use serde::{de::DeserializeOwned, Deserialize, Serialize};
use std::{collections::HashMap, fmt::Display, ops::Index, rc::Rc, string::FromUtf8Error};
use thiserror::Error;
use tracing::{debug, span, Level};

pub static WORLD_KEY: Lazy<EntityKey> = Lazy::new(|| "world".to_string());

pub type EntityKey = String;

pub type ActionArgs<'a> = (&'a Entity, &'a Entity, &'a Entity);

pub type Markdown = markdown::Markdown<Vec<u8>>;

pub fn markdown_to_string(md: Markdown) -> Result<String, FromUtf8Error> {
    String::from_utf8(md.into_inner())
}

type BoxedScope<T> = Box<T>;

pub trait DomainEvent: std::fmt::Debug {}

#[derive(Debug)]
pub struct DomainResult {
    pub events: Vec<Box<dyn DomainEvent>>,
}

pub trait Reply: std::fmt::Debug + erased_serde::Serialize {
    fn to_markdown(&self) -> Result<Markdown>;
}

pub trait InfrastructureFactory: std::fmt::Debug {
    fn new_infrastructure(&self) -> Result<Rc<dyn DomainInfrastructure>>;
}

pub trait DomainInfrastructure: std::fmt::Debug {
    fn ensure_entity(&self, entity_ref: &DynamicEntityRef) -> Result<DynamicEntityRef>;

    fn ensure_optional_entity(
        &self,
        entity_ref: &Option<DynamicEntityRef>,
    ) -> Result<Option<DynamicEntityRef>> {
        match entity_ref {
            Some(e) => Ok(Some(self.ensure_entity(&e)?)),
            None => Ok(None),
        }
    }
}

pub trait PrepareWithInfrastructure {
    fn prepare_with(&mut self, infra: &dyn DomainInfrastructure) -> Result<()>;
}

pub trait Action: std::fmt::Debug {
    fn perform(&self, args: ActionArgs) -> Result<Box<dyn Reply>>;
}

pub trait Scope: PrepareWithInfrastructure + DeserializeOwned {
    fn scope_key() -> &'static str
    where
        Self: Sized;
}

pub trait PrepareEntityByKey {
    fn prepare_entity_by_key<T: Fn(&mut Entity) -> Result<()>>(
        &self,
        key: &EntityKey,
        prepare: T,
    ) -> Result<&Entity>;
}

pub trait LoadEntityByKey {
    fn load_entity_by_key(&self, key: &EntityKey) -> Result<&Entity>;

    fn load_entity_by_ref(&self, entity_ref: &EntityRef) -> Result<&Entity> {
        self.load_entity_by_key(&entity_ref.key)
    }

    fn load_entities_by_refs(&self, entity_refs: Vec<EntityRef>) -> Result<Vec<&Entity>> {
        entity_refs
            .into_iter()
            .map(|re| -> Result<&Entity> { self.load_entity_by_ref(&re) })
            .collect()
    }

    fn load_entities_by_keys(&self, entity_keys: Vec<EntityKey>) -> Result<Vec<&Entity>> {
        entity_keys
            .into_iter()
            .map(|key| -> Result<&Entity> { self.load_entity_by_key(&key) })
            .collect()
    }
}

#[derive(Debug, Serialize)]
pub enum SimpleReply {
    Done,
}

impl Reply for SimpleReply {
    fn to_markdown(&self) -> Result<Markdown> {
        let mut md = Markdown::new(Vec::new());
        md.write("ok!")?;
        Ok(md)
    }
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub enum Item {
    Named(String),
    // ImplicitlyUnheld(String),
    // ImplicitlyHeld(String),
    // ImplicitlyNavigable(String),
}

#[derive(Error, Debug)]
pub enum DomainError {
    #[error("no such scope '{:?}' on entity '{:?}'", .0, .1)]
    NoSuchScope(EntityKey, String),
    #[error("parse failed")]
    ParseFailed(#[source] serde_json::Error),
    #[error("dangling entity")]
    DanglingEntity,
    #[error("anyhow")]
    Anyhow(#[source] anyhow::Error),
    #[error("no infrastructure")]
    NoInfrastructure,
}

impl From<serde_json::Error> for DomainError {
    fn from(source: serde_json::Error) -> Self {
        DomainError::ParseFailed(source)
    }
}

impl From<anyhow::Error> for DomainError {
    fn from(source: anyhow::Error) -> Self {
        DomainError::Anyhow(source)
    }
}

#[derive(Error, Debug)]
pub enum EvaluationError {
    #[error("parse failed")]
    ParseFailed,
}

impl From<nom::Err<nom::error::Error<&str>>> for EvaluationError {
    fn from(source: nom::Err<nom::error::Error<&str>>) -> EvaluationError {
        match source {
            nom::Err::Incomplete(_) => EvaluationError::ParseFailed,
            nom::Err::Error(_e) => EvaluationError::ParseFailed,
            nom::Err::Failure(_e) => EvaluationError::ParseFailed,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EntityRef {
    #[serde(alias = "py/object")]
    py_object: String,
    #[serde(alias = "py/ref")]
    py_ref: String,
    key: EntityKey,
    klass: String,
    name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Identity {
    #[serde(alias = "py/object")]
    py_object: String,
    private: String,
    public: String,
    signature: Option<String>, // TODO Why does this happen in the model?
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Kind {
    #[serde(alias = "py/object")]
    py_object: String,
    identity: Identity,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EntityClass {
    #[serde(alias = "py/type")]
    py_type: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AclRule {
    #[serde(alias = "py/object")]
    py_object: String,
    keys: Vec<String>,
    perm: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Acls {
    #[serde(alias = "py/object")]
    py_object: String,
    rules: Vec<AclRule>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Version {
    #[serde(alias = "py/object")]
    py_object: String,
    i: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Property {
    value: serde_json::Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Props {
    map: HashMap<String, Property>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Entity {
    #[serde(alias = "py/object")]
    py_object: String,
    pub key: String,
    version: Version,
    parent: Option<EntityRef>,
    creator: Option<EntityRef>,
    identity: Identity,
    #[serde(alias = "klass")]
    class: EntityClass,
    acls: Acls,
    props: Props,
    scopes: HashMap<String, serde_json::Value>,

    #[serde(skip)] // Very private
    infra: Option<Rc<dyn DomainInfrastructure>>,
}

impl Display for Entity {
    fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
        fmt.debug_struct(&self.class.py_type)
            .field("key", &self.key)
            .field("name", &self.name())
            .finish()
    }
}

impl Entity {
    pub fn set_infra(&mut self, infra: Rc<dyn DomainInfrastructure>) {
        self.infra = Some(infra);
    }

    fn property_named(&self, name: &str) -> Option<&Property> {
        if self.props.map.contains_key(name) {
            return Some(self.props.map.index(name));
        }
        None
    }

    pub fn name(&self) -> Option<String> {
        if let Some(property) = self.property_named("name") {
            match &property.value {
                serde_json::Value::String(v) => Some(v.to_string()),
                _ => None,
            }
        } else {
            None
        }
    }

    pub fn has_scope<T: Scope>(&self) -> bool {
        self.scopes.contains_key(<T as Scope>::scope_key())
    }

    pub fn scope<T: Scope>(&self) -> Result<BoxedScope<T>, DomainError> {
        let scope_key = <T as Scope>::scope_key();

        let _load_scope_span =
            span!(Level::DEBUG, "scope", key = self.key, scope = scope_key).entered();

        if !self.scopes.contains_key(scope_key) {
            return Err(DomainError::NoSuchScope(
                self.key.clone(),
                scope_key.to_string(),
            ));
        }

        let data = &self.scopes[scope_key];

        debug!("parsing");

        // The call to serde_json::from_value requires owned data and we have a
        // reference to somebody else's. Presumuably so that we don't couple the
        // lifetime of the returned object to the lifetime of the data being
        // referenced? What's the right solution here?
        // Should the 'un-parsed' Scope also owned the parsed data?
        let owned_value = data.clone();
        let mut scope: Box<T> = serde_json::from_value(owned_value)?;

        let _prepare_span = span!(Level::DEBUG, "prepare").entered();

        if let Some(infra) = &self.infra {
            scope.prepare_with(infra.as_ref())?;
        } else {
            return Err(DomainError::NoInfrastructure);
        }

        // Ok, great!
        Ok(scope)
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum DynamicEntityRef {
    RefOnly {
        #[serde(alias = "py/object")]
        py_object: String,
        #[serde(alias = "py/ref")]
        py_ref: String,
        key: String,
        klass: String,
        name: String,
    },
    Entity(Box<Entity>),
}

impl DynamicEntityRef {
    pub fn key(&self) -> &String {
        match self {
            DynamicEntityRef::RefOnly {
                py_object: _,
                py_ref: _,
                key,
                klass: _,
                name: _,
            } => key,
            DynamicEntityRef::Entity(e) => &e.key,
        }
    }
}

impl TryFrom<DynamicEntityRef> for Box<Entity> {
    type Error = DomainError;

    fn try_from(value: DynamicEntityRef) -> Result<Self, Self::Error> {
        match value {
            DynamicEntityRef::RefOnly {
                py_object: _,
                py_ref: _,
                key: _,
                klass: _,
                name: _,
            } => Err(DomainError::DanglingEntity),
            DynamicEntityRef::Entity(e) => Ok(e),
        }
    }
}

impl From<DynamicEntityRef> for EntityRef {
    fn from(value: DynamicEntityRef) -> Self {
        match value {
            DynamicEntityRef::RefOnly {
                py_object,
                py_ref,
                key,
                klass,
                name,
            } => EntityRef {
                py_object,
                py_ref,
                key,
                klass,
                name,
            },
            DynamicEntityRef::Entity(e) => e.as_ref().into(),
        }
    }
}

impl From<&Entity> for EntityRef {
    fn from(e: &Entity) -> Self {
        EntityRef {
            py_object: "todo!".to_string(),
            py_ref: "todo!".to_string(),
            key: e.key.to_string(),
            klass: "todo!".to_string(),
            name: e.name().unwrap_or_default(),
        }
    }
}

#[derive(Debug)]
pub struct PersistedEntity {
    pub key: String,
    pub gid: u32,
    pub version: u32,
    pub serialized: String,
}
