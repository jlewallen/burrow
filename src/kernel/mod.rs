pub mod english;
pub mod infra;
pub mod model;
pub mod scopes;

pub use english::*;
pub use infra::*;
pub use model::*;
pub use scopes::*;

pub use replies::*;

pub type ReplyResult = anyhow::Result<Box<dyn Reply>>;
