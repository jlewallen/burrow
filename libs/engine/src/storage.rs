use anyhow::Result;
use std::{
    collections::HashMap,
    rc::Rc,
    sync::{Arc, RwLock},
};

use kernel::{EntityGid, EntityKey, LookupBy};

pub trait EntityStorage {
    fn load(&self, lookup: &LookupBy) -> Result<Option<PersistedEntity>>;
    fn save(&self, entity: &PersistedEntity) -> Result<()>;
    fn delete(&self, entity: &PersistedEntity) -> Result<()>;
    fn begin(&self) -> Result<()>;
    fn rollback(&self, benign: bool) -> Result<()>;
    fn commit(&self) -> Result<()>;
    fn query_all(&self) -> Result<Vec<PersistedEntity>>;
}

pub trait EntityStorageFactory: Send + Sync {
    fn create_storage(&self) -> Result<Rc<dyn EntityStorage>>;
}

#[derive(Clone, Debug)]
pub struct PersistedEntity {
    pub key: String,
    pub gid: u64,
    pub version: u64,
    pub serialized: String,
}

impl PersistedEntity {
    pub fn to_json_value(&self) -> Result<serde_json::Value> {
        Ok(serde_json::from_str(&self.serialized)?)
    }
}

#[derive(Default)]
pub struct InMemoryEntityStorageFactory {
    entities: Arc<RwLock<HashMap<EntityKey, PersistedEntity>>>,
}

impl EntityStorageFactory for InMemoryEntityStorageFactory {
    fn create_storage(&self) -> Result<Rc<dyn EntityStorage>> {
        Ok(Rc::new(InMemoryEntityStorage {
            entities: self.entities.clone(),
            pending: Default::default(),
        }))
    }
}

enum Pending {
    Save(PersistedEntity),
    Delete(PersistedEntity),
}

pub struct InMemoryEntityStorage {
    entities: Arc<RwLock<HashMap<EntityKey, PersistedEntity>>>,
    pending: RwLock<Vec<Pending>>,
}

impl EntityStorage for InMemoryEntityStorage {
    fn load(&self, lookup: &LookupBy) -> Result<Option<PersistedEntity>> {
        let entities = self.entities.read().expect("Lock error");
        let entity = entities
            .iter()
            .filter(|(_, e)| match lookup {
                LookupBy::Key(key) => e.key == key.key_to_string(),
                LookupBy::Gid(gid) => EntityGid::new(e.gid) == **gid,
            })
            .map(|(_, e)| e)
            .next();

        Ok(entity.cloned())
    }

    fn save(&self, entity: &PersistedEntity) -> Result<()> {
        let mut pending = self.pending.write().expect("Lock error");
        pending.push(Pending::Save(entity.clone()));

        Ok(())
    }

    fn delete(&self, entity: &PersistedEntity) -> Result<()> {
        let mut pending = self.pending.write().expect("Lock error");
        pending.push(Pending::Delete(entity.clone()));

        Ok(())
    }

    fn begin(&self) -> Result<()> {
        let mut pending = self.pending.write().expect("Lock error");
        pending.clear();

        Ok(())
    }

    fn rollback(&self, _benign: bool) -> Result<()> {
        let mut pending = self.pending.write().expect("Lock error");
        pending.clear();

        Ok(())
    }

    fn commit(&self) -> Result<()> {
        let mut pending = self.pending.write().expect("Lock error");
        let mut entities = self.entities.write().expect("Lock error");

        for pending in pending.iter() {
            match pending {
                Pending::Save(e) => entities.insert(EntityKey::new(&e.key), e.clone()),
                Pending::Delete(e) => entities.remove(&EntityKey::new(&e.key)),
            };
        }

        pending.clear();

        Ok(())
    }

    fn query_all(&self) -> Result<Vec<PersistedEntity>> {
        let entities = self.entities.read().expect("Lock error");

        Ok(entities.values().map(|e| e.clone()).collect())
    }
}
