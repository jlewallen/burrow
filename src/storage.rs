use crate::kernel::EntityKey;
use anyhow::Result;
use std::rc::Rc;
use tracing::debug;

pub trait EntityStorage {
    fn load(&self, key: &EntityKey) -> Result<PersistedEntity>;
    fn save(&self, entity: &PersistedEntity) -> Result<()>;
    fn begin(&self) -> Result<()>;
    fn rollback(&self) -> Result<()>;
    fn commit(&self) -> Result<()>;
}

pub trait EntityStorageFactory {
    fn create_storage(&self) -> Result<Rc<dyn EntityStorage>>;
}

#[derive(Debug)]
pub struct PersistedEntity {
    pub key: String,
    pub gid: u64,
    pub version: u64,
    pub serialized: String,
}

pub mod sqlite {
    use super::*;
    use anyhow::anyhow;
    use rusqlite::Connection;
    use tracing::info;

    pub struct SqliteStorage {
        conn: Connection,
    }

    impl SqliteStorage {
        pub fn new(path: &str) -> Result<Self> {
            let conn = if path == ":memory:" {
                Connection::open_in_memory()?
            } else {
                Connection::open(&path)?
            };

            {
                let mut stmt = conn.prepare(
                    r#"
                CREATE TABLE IF NOT EXISTS entities (
                    key TEXT NOT NULL PRIMARY KEY,
                    version INTEGER NOT NULL,
                    gid INTEGER,
                    serialized TEXT NOT NULL
                )"#,
                )?;
                let _ = stmt.execute([])?;
            }

            Ok(SqliteStorage { conn: conn })
        }
    }

    impl EntityStorage for SqliteStorage {
        fn load(&self, key: &EntityKey) -> Result<PersistedEntity> {
            let mut stmt = self
                .conn
                .prepare("SELECT key, gid, version, serialized FROM entities WHERE key = ?;")?;

            debug!("querying");

            let mut entities = stmt.query_map([key.key_to_string()], |row| {
                Ok(PersistedEntity {
                    key: row.get(0)?,
                    gid: row.get(1)?,
                    version: row.get(2)?,
                    serialized: row.get(3)?,
                })
            })?;

            match entities.next() {
                Some(p) => Ok(p?),
                _ => Err(anyhow!("entity with key '{}' not found", key)),
            }
        }

        fn save(&self, entity: &PersistedEntity) -> Result<()> {
            let affected = if entity.version == 1 {
                let mut stmt = self.conn.prepare(
                    "INSERT INTO entities (key, gid, version, serialized) VALUES (?1, ?2, ?3, ?4)",
                )?;

                stmt.execute((
                    &entity.key.to_string(),
                    &entity.gid,
                    &entity.version,
                    &entity.serialized,
                ))?
            } else {
                let mut stmt = self.conn.prepare(
                    "UPDATE entities SET gid = ?1, version = ?2, serialized = ?3 WHERE key = ?4 AND version = ?5",
                )?;

                stmt.execute((
                    &entity.gid,
                    &entity.version,
                    &entity.serialized,
                    &entity.key.to_string(),
                    &entity.version - 1,
                ))?
            };

            if affected != 1 {
                Err(anyhow!("no rows affected by save"))
            } else {
                Ok(())
            }
        }

        fn begin(&self) -> Result<()> {
            info!("tx:begin");

            self.conn.execute("BEGIN TRANSACTION", [])?;

            Ok(())
        }

        fn rollback(&self) -> Result<()> {
            info!("tx:rollback");

            self.conn.execute("ROLLBACK TRANSACTION", [])?;

            Ok(())
        }

        fn commit(&self) -> Result<()> {
            info!("tx:commit");

            self.conn.execute("COMMIT TRANSACTION", [])?;

            Ok(())
        }
    }

    pub struct Factory {
        path: String,
    }

    impl Factory {
        pub fn new(path: &str) -> Result<Box<Factory>> {
            Ok(Box::new(Factory {
                path: path.to_string(),
            }))
        }
    }

    impl EntityStorageFactory for Factory {
        fn create_storage(&self) -> Result<Rc<dyn EntityStorage>> {
            Ok(Rc::new(SqliteStorage::new(&self.path)?))
        }
    }

    #[cfg(test)]
    mod tests {
        use super::*;
        use anyhow::Result;

        fn get_storage() -> Result<Rc<dyn EntityStorage>> {
            let s = Factory::new(":memory:")?;

            Ok(s.create_storage()?)
        }

        #[test]
        fn it_queries_for_entity_by_missing_key() -> Result<()> {
            let s = get_storage()?;

            assert!(s.load(&EntityKey::new("world")).is_err());

            Ok(())
        }

        #[test]
        fn it_inserts_a_new_entity() -> Result<()> {
            let s = get_storage()?;

            s.save(&PersistedEntity {
                key: "world".to_string(),
                gid: 1,
                version: 1,
                serialized: "{}".to_string(),
            })
        }

        #[test]
        fn it_queries_for_entity_by_key() -> Result<()> {
            let s = get_storage()?;

            s.save(&PersistedEntity {
                key: "world".to_string(),
                gid: 1,
                version: 1,
                serialized: "{}".to_string(),
            })?;

            s.load(&EntityKey::new("world"))?;

            Ok(())
        }

        #[test]
        fn it_updates_an_existing_new_entity() -> Result<()> {
            let s = get_storage()?;

            s.save(&PersistedEntity {
                key: "world".to_string(),
                gid: 1,
                version: 1,
                serialized: "{}".to_string(),
            })?;

            let mut p1 = s.load(&EntityKey::new("world"))?;

            assert_eq!(1, p1.version);

            p1.version += 1;

            s.save(&p1)?;

            let p2 = s.load(&EntityKey::new("world"))?;

            assert_eq!(2, p2.version);

            Ok(())
        }

        #[test]
        fn it_inserts_a_new_entity_in_a_rolled_back_transaction_inserts_nothing() -> Result<()> {
            let s = get_storage()?;

            s.begin()?;

            s.save(&PersistedEntity {
                key: "world".to_string(),
                gid: 1,
                version: 1,
                serialized: "{}".to_string(),
            })?;

            s.rollback()?;

            assert!(s.load(&EntityKey::new("world")).is_err());

            Ok(())
        }

        #[test]
        fn it_inserts_a_new_entity_in_a_committed_transaction() -> Result<()> {
            let s = get_storage()?;

            s.begin()?;

            s.save(&PersistedEntity {
                key: "world".to_string(),
                gid: 1,
                version: 1,
                serialized: "{}".to_string(),
            })?;

            s.commit()?;

            s.load(&EntityKey::new("world"))?;

            Ok(())
        }
    }
}
