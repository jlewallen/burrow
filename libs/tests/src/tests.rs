use anyhow::Result;
use tokio::task::JoinHandle;

use crate::{evaluate_fixture, test_domain_with, HoldingKeyInVessel, Noop, WorldFixture, USERNAME};
use engine::storage::{PersistedEntity, StorageFactory};
use engine::{Domain, Session, SessionOpener};
use kernel::JsonValue;

async fn test_domain() -> Result<AsyncFriendlyDomain> {
    let storage_factory = sqlite::Factory::new(sqlite::MEMORY_SPECIAL)?;

    storage_factory.migrate()?;

    Ok(AsyncFriendlyDomain::wrap(test_domain_with(
        storage_factory,
    )?))
}

#[tokio::test]
async fn it_evaluates_a_simple_look() -> Result<()> {
    let domain = test_domain().await?;
    domain.evaluate::<HoldingKeyInVessel>(&["look"]).await?;
    insta::assert_json_snapshot!(domain.snapshot().await?);
    domain.stop().await?;

    Ok(())
}

#[tokio::test]
async fn it_evaluates_two_simple_looks_same_session() -> Result<()> {
    let domain = test_domain().await?;
    domain
        .evaluate::<HoldingKeyInVessel>(&["look", "look"])
        .await?;
    insta::assert_json_snapshot!(domain.snapshot().await?);
    domain.stop().await?;

    Ok(())
}

#[tokio::test]
async fn it_evaluates_two_simple_looks_separate_session() -> Result<()> {
    let domain = test_domain().await?;
    domain.evaluate::<HoldingKeyInVessel>(&["look"]).await?;
    domain.evaluate::<Noop>(&["look"]).await?;
    insta::assert_json_snapshot!(domain.snapshot().await?);
    domain.stop().await?;

    Ok(())
}

#[tokio::test]
async fn it_can_drop_held_container() -> Result<()> {
    let domain = test_domain().await?;
    domain.evaluate::<HoldingKeyInVessel>(&["look"]).await?;
    domain.evaluate::<Noop>(&["drop vessel"]).await?;
    insta::assert_json_snapshot!(domain.snapshot().await?);
    domain.stop().await?;

    Ok(())
}

#[tokio::test]
async fn it_can_rehold_dropped_container() -> Result<()> {
    let domain = test_domain().await?;
    domain.evaluate::<HoldingKeyInVessel>(&["look"]).await?;
    domain.evaluate::<Noop>(&["drop vessel"]).await?;
    domain.evaluate::<Noop>(&["hold vessel"]).await?;
    insta::assert_json_snapshot!(domain.snapshot().await?);
    domain.stop().await?;

    Ok(())
}

#[tokio::test]
async fn it_can_go_east() -> Result<()> {
    let domain = test_domain().await?;
    domain.evaluate::<HoldingKeyInVessel>(&["look"]).await?;
    domain.evaluate::<Noop>(&["go east"]).await?;
    domain.evaluate::<Noop>(&["look"]).await?;
    insta::assert_json_snapshot!(domain.snapshot().await?);
    domain.stop().await?;

    Ok(())
}

/*
#[cfg(test)]
#[ctor::ctor]
fn initialize_tests() {
    plugins_core::log_test();
}
*/

#[derive(Clone)]
pub struct AsyncFriendlyDomain {
    domain: Domain,
}

impl AsyncFriendlyDomain {
    pub fn wrap(domain: Domain) -> Self {
        Self { domain }
    }

    pub async fn query_all(&self) -> Result<Vec<PersistedEntity>> {
        self.domain.query_all()
    }

    #[cfg(test)]
    pub async fn snapshot(&self) -> Result<JsonValue> {
        let json: Vec<JsonValue> = self
            .query_all()
            .await?
            .into_iter()
            .map(|p| p.to_json_value())
            .collect::<Result<_>>()?;

        Ok(JsonValue::Array(json))
    }

    pub async fn evaluate<W>(&self, text: &'static [&'static str]) -> Result<()>
    where
        W: WorldFixture + Default,
    {
        let handle: JoinHandle<Result<()>> = tokio::task::spawn_blocking({
            let sessions = self.clone();
            move || {
                evaluate_fixture::<W, _>(&sessions, USERNAME, text)?;

                Ok(())
            }
        });

        Ok(handle.await??)
    }

    pub async fn stop(&self) -> Result<()> {
        let domain = self.domain.clone();
        tokio::task::spawn_blocking(move || domain.stop()).await?
    }
}

impl SessionOpener for AsyncFriendlyDomain {
    fn open_session(&self) -> Result<std::rc::Rc<Session>> {
        self.domain.open_session()
    }
}
