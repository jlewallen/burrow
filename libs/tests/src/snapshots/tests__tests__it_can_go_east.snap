---
source: libs/tests/src/tests.rs
expression: domain.snapshot().await?
---
[
  {
    "acls": {
      "rules": []
    },
    "class": {
      "name": "scopes.ItemClass"
    },
    "creator": null,
    "identity": {
      "private": "Private#0",
      "public": "Public#0"
    },
    "key": "world",
    "parent": null,
    "scopes": {
      "identifiers": {
        "acls": {
          "rules": []
        },
        "gid": 5
      },
      "props": {
        "core": {
          "gid": {
            "acls": null,
            "value": 0
          },
          "name": {
            "acls": null,
            "value": "World"
          }
        }
      },
      "usernames": {
        "users": {
          "burrow": "E-3"
        }
      }
    }
  },
  {
    "acls": {
      "rules": []
    },
    "class": {
      "name": "scopes.ItemClass"
    },
    "creator": null,
    "identity": {
      "private": "Private#1",
      "public": "Public#1"
    },
    "key": "E-0",
    "parent": null,
    "scopes": {
      "occupyable": {
        "acls": {
          "rules": []
        },
        "occupancy": 0,
        "occupied": [
          {
            "class": "scopes.ItemClass",
            "gid": 4,
            "key": "E-3",
            "name": "Living"
          }
        ],
        "routes": null
      },
      "props": {
        "core": {
          "gid": {
            "acls": null,
            "value": 1
          },
          "name": {
            "acls": null,
            "value": "Place"
          }
        }
      }
    }
  },
  {
    "acls": {
      "rules": []
    },
    "class": {
      "name": "scopes.ItemClass"
    },
    "creator": null,
    "identity": {
      "private": "Private#2",
      "public": "Public#2"
    },
    "key": "E-1",
    "parent": null,
    "scopes": {
      "carryable": {
        "kind": {
          "identity": {
            "private": "Private#3",
            "public": "Public#3"
          }
        },
        "quantity": 1.0
      },
      "location": {
        "container": {
          "class": "scopes.ItemClass",
          "gid": 3,
          "key": "E-2",
          "name": "Vessel"
        }
      },
      "props": {
        "core": {
          "gid": {
            "acls": null,
            "value": 2
          },
          "name": {
            "acls": null,
            "value": "Key"
          }
        }
      }
    }
  },
  {
    "acls": {
      "rules": []
    },
    "class": {
      "name": "scopes.ItemClass"
    },
    "creator": null,
    "identity": {
      "private": "Private#4",
      "public": "Public#4"
    },
    "key": "E-2",
    "parent": null,
    "scopes": {
      "carryable": {
        "kind": {
          "identity": {
            "private": "Private#5",
            "public": "Public#5"
          }
        },
        "quantity": 1.0
      },
      "containing": {
        "capacity": null,
        "holding": [
          {
            "class": "scopes.ItemClass",
            "gid": 2,
            "key": "E-1",
            "name": "Key"
          }
        ],
        "produces": {}
      },
      "location": {
        "container": {
          "class": "scopes.ItemClass",
          "gid": 4,
          "key": "E-3",
          "name": "Living"
        }
      },
      "props": {
        "core": {
          "gid": {
            "acls": null,
            "value": 3
          },
          "name": {
            "acls": null,
            "value": "Vessel"
          }
        }
      }
    }
  },
  {
    "acls": {
      "rules": []
    },
    "class": {
      "name": "scopes.ItemClass"
    },
    "creator": null,
    "identity": {
      "private": "Private#6",
      "public": "Public#6"
    },
    "key": "E-3",
    "parent": null,
    "scopes": {
      "containing": {
        "capacity": null,
        "holding": [
          {
            "class": "scopes.ItemClass",
            "gid": 3,
            "key": "E-2",
            "name": "Vessel"
          }
        ],
        "produces": {}
      },
      "occupying": {
        "area": {
          "class": "scopes.ItemClass",
          "gid": 1,
          "key": "E-0",
          "name": "Place"
        }
      },
      "props": {
        "core": {
          "gid": {
            "acls": null,
            "value": 4
          },
          "name": {
            "acls": null,
            "value": "Living"
          }
        }
      },
      "wearing": {
        "wearing": []
      }
    }
  },
  {
    "acls": {
      "rules": []
    },
    "class": {
      "name": "scopes.ItemClass"
    },
    "creator": null,
    "identity": {
      "private": "Private#7",
      "public": "Public#7"
    },
    "key": "E-4",
    "parent": null,
    "scopes": {
      "containing": {
        "capacity": null,
        "holding": [],
        "produces": {}
      },
      "occupyable": {
        "acls": {
          "rules": []
        },
        "occupancy": 0,
        "occupied": [],
        "routes": [
          {
            "simple": {
              "name": "East",
              "to": {
                "class": "scopes.ItemClass",
                "gid": 1,
                "key": "E-0",
                "name": "Place"
              }
            }
          }
        ]
      },
      "props": {
        "core": {
          "gid": {
            "acls": null,
            "value": 5
          },
          "name": {
            "acls": null,
            "value": "Welcome Area"
          }
        }
      }
    }
  }
]
